keras_nlp/metrics/bleu.py

```diff
@@ -0,0 +1,394 @@
+        def validate_and_fix_rank(inputs, tensor_name, base_rank=0):
+            if not isinstance(inputs, (tf.Tensor, tf.RaggedTensor)):
+                inputs = tf.convert_to_tensor(inputs)
+
+            if inputs.shape.rank == base_rank:
+                return inputs[tf.newaxis]
+            elif inputs.shape.rank == base_rank + 1:
+                return inputs
+            elif inputs.shape.rank == base_rank + 2:
+                if tf.shape(inputs)[-1] != 1:
+                    raise ValueError(
+                        f"{tensor_name} is of rank {input.shape.rank}. The "
+                        f"last dimension must be of size 1."
+                    )
+                return tf.squeeze(inputs, axis=-1)
+            else:
+                raise ValueError(
+                    f"{tensor_name} must be of rank {base_rank}, {base_rank+1} "
+                    f"or {base_rank+2}. Found rank: {inputs.shape.rank}"
+                )
+
+        def calculate_bleu_score(references, translation):
+            if references.dtype == tf.string:
+                references = tensor_to_string_list(references)
+                translation = tensor_to_string_list(translation)
+            else:
+                references = tensor_to_list(references)
+                translation = tensor_to_list(translation)
+
+            matches = self._matches.numpy().tolist()
+            possible_matches = self._possible_matches.numpy().tolist()
+            translation_length = self._translation_length.numpy()
+            reference_length = self._reference_length.numpy()
+
+            (
+                bleu_score,
+                matches,
+                possible_matches,
+                translation_length,
+                reference_length,
+            ) = self._corpus_bleu(
+                reference_corpus=references,
+                translation_corpus=translation,
+                matches_by_order=matches,
+                possible_matches_by_order=possible_matches,
+                translation_length=translation_length,
+                reference_length=reference_length,
+                max_order=self.max_order,
+                smooth=self.smooth,
+            )
+            return (
+                tf.constant(bleu_score, dtype=self.dtype),
+                tf.constant(matches, dtype=self.dtype),
+                tf.constant(possible_matches, dtype=self.dtype),
+                tf.constant(translation_length, dtype=self.dtype),
+                tf.constant(reference_length, dtype=self.dtype),
+            )
+
+        y_true = validate_and_fix_rank(y_true, "y_true", 1)
+        y_pred = validate_and_fix_rank(y_pred, "y_pred", 0)
+
+        # Tokenize the inputs.
+        y_true = self.tokenizer(y_true)
+        y_pred = self.tokenizer(y_pred)
+
+        (
+            bleu_score,
+            matches,
+            possible_matches,
+            translation_length,
+            reference_length,
+        ) = tf.py_function(
```

Comment:
given that you are converting everything to numpy, would this do better as a numpy_function call?

Reply:
sg, let's stick with py_function for now